xtask/src/commands/release.rs

```diff
@@ -197,6 +211,31 @@ impl Prepare {
                 "--package",
                 "apollo-router"
             ]),
+            Version::Nightly => {
+                let head_commit: String = match Repository::open_from_env() {
+                    Ok(repo) => {
+                        let revspec = repo.revparse("HEAD")?;
+                        if revspec.mode().contains(git2::RevparseMode::SINGLE) {
+                            let mut full_hash = revspec.from().unwrap().id().to_string();
+                            full_hash.truncate(8);
+                            full_hash
+                        } else {
+                            panic!("unexpected rev-parse HEAD");
+                        }
+                    }
+                    Err(e) => panic!("failed to open: {}", e),
+                };
+
+                replace_in_file!(
+                    "./apollo-router/Cargo.toml",
+                    r#"^(?P<existingVersion>version\s*=\s*)"[^"]+""#,
+                    format!(
+                        "${{existingVersion}}\"0.0.0-nightly.{}+{}\"",
+                        Utc::now().format("%Y%m%d"),
+                        head_commit
+                    )
+                );
```

Comment:
This deviates from the pattern used in the other arms of the `match` in the way which it assigns the `version`.  It does that because [`cargo set-version` from `cargo edit`](https://github.com/killercup/cargo-edit#cargo-set-version) doesn't ever let you set the version lower than the current `version`.

I considered an approach where we instead stuck the `nightly` on the existing `version` number, but it became fragile depending on the `preid` (e.g., `alpha`, `beta`) and it seemed most reasonable (even in terms of defining breaking changes in the spirit of semantic versioning) that `0.0.0` was a more sensical number to use on nightlies which seemed compatible with tooling (and _just_ thought provoking enough to humans for them to consider its meaning).